src/ConvexMethod.cpp

```diff
@@ -22,47 +52,204 @@ std::optional<LeftRightResults> ConvexMethod::classify(const geometry_msgs::msg:
     return classification;
 }
 
-std::vector<cv::Point> ConvexMethod::get_detections_vector(const geometry_msgs::msg::PoseArray& detections_2d) {
-    //TODO impl
-    (void)detections_2d;
-    return {};
+std::vector<cv::Point2d> ConvexMethod::get_detections_vector(const geometry_msgs::msg::PoseArray& detections_array) {
+    std::vector<cv::Point2d> detections_2d{};
+    for (const auto& pose : detections_array.poses) {
+        detections_2d.emplace_back(pose.position.x, pose.position.y);
+    }
+    return detections_2d;
 }
 
-std::vector<cv::Point> ConvexMethod::get_convex_hull() {
-    //TODO impl
-    return {};
+std::vector<cv::Point2d> ConvexMethod::get_convex_hull(const std::vector<cv::Point2d>& detections_2d) {
+    std::vector<cv::Point2d> convex_hull{};
+    cv::convexHull(detections_2d, convex_hull, false);
+    return convex_hull;
 }
 
-bool ConvexMethod::is_convex_hull_valid() {
-    //TODO impl
-    return false;
+bool ConvexMethod::is_convex_hull_valid(const std::vector<cv::Point2d>& convex_hull) {
+    bool is_valid{false};
+    double area = 0.0;
+    std::size_t num_vertices{convex_hull.size()};
+    for (std::size_t i{0}; i <= num_vertices; ++i) {
+        std::size_t j{(i + 1U) % (num_vertices + 1)};
+        const cv::Point2d& first = convex_hull[i];
+        const cv::Point2d& second = convex_hull[j];
+        area += 0.5 * (first.x * second.y - second.x * first.y);
+    }
+
+    if (std::fabs(area) > params.convex_hull_area_threshold) {
+        is_valid = true;
+    }
+
+    return is_valid;
 }
 
-Scenario ConvexMethod::determine_scenario() {
+Scenario ConvexMethod::determine_scenario(const std::vector<cv::Point2d>& detections_2d) {
     //TODO impl
-    return Scenario::kStraight;
+    Scenario scenario{Scenario::kStraight};
+
+    if (detections_2d.size() > 0) {
+        const double sum_in_y =
+            std::accumulate(std::begin(detections_2d), std::end(detections_2d), 0.0,
+                            [](const auto& accumulation, const auto& point) { return accumulation + point.y; });
+        const double center_of_mass_in_y{sum_in_y / detections_2d.size()};
+        if (center_of_mass_in_y > params.turn_threshold) {
+            scenario = Scenario::kLeft;
+        } else if (center_of_mass_in_y < -params.turn_threshold) {
+            scenario = Scenario::kRight;
+        }
+    }
+    return scenario;
 }
 
-LeftRightResults StraightScenarioClassifier::classify(const std::vector<cv::Point>& convex_hull,
-                                                      const std::vector<cv::Point>& detections_2d) {
-    //TODO impl
-    (void)convex_hull;
-    (void)detections_2d;
-    return {};
+LeftRightResults StraightScenarioClassifier::classify(const std::vector<cv::Point2d>& convex_hull,
+                                                      const std::vector<cv::Point2d>& detections_2d) {
+    (void)convex_hull;  // Unused but needed to satisfy interface. This hurts me deeply
```

Comment:
Whats the reasoning for this? Never seen this before.

Reply:
Thanks for making this better!